atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/SchemaMutationLock.java

```diff
@@ -68,18 +66,102 @@ public SchemaMutationLock(
-        long lockId = waitForSchemaMutationLock();
+    private class HeartbeatThread {
+        private final Thread heartbeatThread;
+        private final long lockId;
+        private volatile boolean keepBeating;
+        private volatile long heartbeatCount;
+
+        HeartbeatThread(long lockId) {
+            this.heartbeatThread = new Thread(this::heartbeat);
+            this.lockId = lockId;
+            this.heartbeatCount = 0;
+        }
+
+        void startBeating() {
+            keepBeating = true;
+            heartbeatThread.start();
+        }
+
+        void stopBeating() {
+            keepBeating = false;
+            try {
+                // make sure beating's stopped
+                heartbeatThread.join();
+            } catch (InterruptedException e) {
+                LOGGER.debug("Interrupted while trying to stop the heartbeat heartbeatThread.");
+            }
+        }
+
+        void heartbeat() {
+            while (keepBeating) {
+                try {
+                    clientPool.runWithRetry((FunctionCheckedException<Cassandra.Client, Void, Exception>) client -> {
+                        Column ourUpdate = lockColumnWithStrValue(String.format(
+                                CassandraConstants.GLOBAL_DDL_LOCK_FORMAT, lockId, heartbeatCount + 1));
+
+                        List<Column> expected = ImmutableList.of(lockColumnWithStrValue(String.format(
+                                CassandraConstants.GLOBAL_DDL_LOCK_FORMAT, lockId, heartbeatCount)));
+
+                        CASResult casResult = writeDdlLockWithCas(client, expected, ourUpdate);
+
+                        if (casResult.isSuccess()) {
+                            heartbeatCount++;
+                        } else {
+                            handleForcedLockClear(casResult, lockId, heartbeatCount);
+                        }
+                        return null;
+                    });
+                    Thread.sleep(configManager.getConfig().heartbeatTimePeriodMillis());
+                } catch (Exception e) {
+                    throw Throwables.throwUncheckedException(e);
+                }
+            }
+        }
+    }
 
+    void runWithLock(Action action) {
+        if (!supportsCas) {
+            runWithLockWithoutCas(action);
+            return;
+        }
+
+        long lockId = waitForSchemaMutationLock();
+        heartbeatThread = new HeartbeatThread(lockId);
+        heartbeatThread.startBeating();
         try {
             action.execute();
         } catch (Exception e) {
             throw Throwables.throwUncheckedException(e);
         } finally {
+            heartbeatThread.stopBeating();
             schemaMutationUnlock(lockId);
         }
     }
 
+    private void runWithLockWithoutCas(Action action) {
+        if (configManager.getConfig().servers().size() > 1) {
+            throw new UnsupportedOperationException("Running a clustered service with a version of Cassandra"
```

Comment:
nit: should probably be IllegalStateException, which we could enforce via Preconditions.checkState. The other piece is that we probably need to hit C* directly to determine if node count > 1 because the Atlas config may only have 1 seed node listed, but C* cluster may have N> 1 nodes that they identify via their config & gossip.